examples/mandelbrot_image_gpu.py

```diff
@@ -0,0 +1,285 @@
+        chunks.append(f(*calc_args))
+
+    print(f"{datetime.now()}: finished scheduling")
+
+    img = Image.new("L", (size.x, size.y))
+
+    for c in range(0, num_chunks):
+        start_y = c * chunk_size
+        end_y = min(start_y + chunk_size, size.y)
+
+        if end_y <= start_y:
+            break
+
+        chunk = chunks.pop(0)
+
+        if use_ray:
+            chunk = ray.get(chunk)
+
+        chunk_img_size = (size.x, end_y - start_y)
+        box = (0, start_y)
+
+        print(f"{datetime.now()}: got chunk {c}, size: {chunk_img_size}, box: {box}")
+
+        chunk_img = Image.frombytes("L", size=chunk_img_size, data=bytearray(chunk.data))
+        img.paste(chunk_img, box=box)
+
+        print(f"{datetime.now()}: processed chunk {c}")
+
+    img.show()
+
+    current_time_str = datetime.now(tz=timezone.utc).strftime("%Y%m%d_%H%M%S%z")
+    filename = output_file or f"mandel-{size.x}x{size.y}-{current_time_str}.png"
+    img.save(filename, "PNG")
+
+    print(f"{datetime.now()}: saved as {filename}")
+
+
+def argument_parser():
+    parser = argparse.ArgumentParser("mandelbrot on ray")
+    parser.add_argument(
+        "-s",
+        "--size",
+        nargs=2,
+        metavar=("X", "Y"),
+        help="size of the output image, default=%(default)s",
+        type=int,
+        default=(500, 500),
+    )
+    parser.add_argument(
+        "-c",
+        "--center",
+        nargs=2,
+        metavar=("X", "Y"),
+        help="center of the drawn region, default=%(default)s",
+        type=float,
+        default=(-0.743643135, 0.131825963),
+    )
+    parser.add_argument(
+        "-z",
+        "--zoom",
+        help="magnification of the drawn region, default=%(default)s",
+        type=float,
+        default=200000,
+    )
+    parser.add_argument(
+        "-i",
+        "--max-iterations",
+        help="maximum number of iterations to perform per pixel, default=%(default)s",
+        type=int,
+        default=500,
+    )
+    parser.add_argument(
+        "-n",
+        "--num-chunks",
+        help="number of chunks to divide output into, default=%(default)s",
+        type=int,
+        default=16,
+    )
+    parser.add_argument(
+        "-f",
+        "--output-file",
+        help="name of file to save output into, defaults to predefined template",
+        type=str,
+        nargs="?",
+    )
+    parser.add_argument(
+        "--ray-num-cpus",
+        help="number of CPUs for ray to use",
+        type=int,
+    )
+    parser.add_argument("-r", "--use-ray", action="store_true", help="use ray")
+    parser.add_argument(
+        "-R", "--no-use-ray", dest="use_ray", action="store_false", help="don't use ray"
+    )
+    parser.add_argument("-g", "--use-gpu", action="store_true", help="use GPU")
+    parser.add_argument(
+        "-G", "--no-use-gpu", dest="use_gpu", action="store_false", help="don't use GPU"
+    )
+    parser.set_defaults(use_ray=True, use_gpu=True)
```

Comment:
If we are passing this argument in multiple functions maybe it would be a good idea to have one `mandelbrot_image.py` example that will cover both CPU and GPU use cases?

Reply:
there are two issues:

* this example requires the `numba` library which needs to be installed on top of `pillow` and obviously, requires either an image that already includes `numba` - or - requires the library to be installed when spawning the cluster
* it's inherently much more complicated that the original example...

so, consequently, I'd prefer to leave the original example in place and have this, more optimized and complicated one as an addition...

Reply:
:+1: marking as resolved